crates/biome_js_semantic/src/events.rs

```diff
@@ -158,44 +152,84 @@ impl SemanticEvent {
 /// ```
 #[derive(Default, Debug)]
 pub struct SemanticEventExtractor {
+    // Event queue
     stash: VecDeque<SemanticEvent>,
+    /// Stack of scopes
     scopes: Vec<Scope>,
-    next_scope_id: usize,
-    /// At any point this is the set of available bindings and
-    /// the range of its declaration
+    // Number of generated scopes
+    // This allows assigning a unique scope id for every scope.
+    scope_count: usize,
+    /// At any point this is the set of available bindings in the current scope
     bindings: FxHashMap<BindingName, BindingInfo>,
+    // Type parameters bound in a `infer T` clause.
     infers: Vec<TsTypeParameterName>,
 }
 
-/// Holds the text range of the token when it is bound,
-/// along with the kind of declaration
-#[derive(Debug)]
-struct BindingInfo {
-    text_range: TextRange,
-    /// For export determination, it is necessary to provide
-    /// information on how a specific token is bound
-    declaration_kind: JsSyntaxKind,
+/// A declaration is either a type, a value, or both.
+///
+/// For example a class is botha  type and a value in TypeScript.
+/// A variable declaration is always a value, and a type alias is always a type.
+#[derive(Debug, Hash, Eq, PartialEq, Copy, Clone)]
+enum DeclarationKind {
+    Type,
+    Value,
+    Both,
 }
 
+/// A binding name is either a type or a value.
+///
+/// If a declaration is both a type and a value,
+/// then two bindings are associated to the declaration a type and a value bindings.
+///
+/// This allows to properly detect type and value shadowing in inner scopes.
 #[derive(Debug, Hash, Eq, PartialEq, Clone)]
-struct BindingName {
-    name: TokenText,
+enum BindingName {
+    Type(TokenText),
+    Value(TokenText),
 }
 
+impl BindingName {
+    /// Returns a binding with the same name, but the other kind.
+    fn dual(self) -> Self {
+        match self {
+            Self::Type(name) => Self::Value(name),
+            Self::Value(name) => Self::Type(name),
+        }
+    }
```

Comment:
I don't understand the purpose of this method. It seems more like a hack.

Reply:
It is only used when matching Export and bindings.

When we encounter an `export` we emit two export references: one for a type binding and the other for a value binding.
To avoid producing an unresolved reference when only the type or the value is exported, we check if it exists a dual binding. If it exists, then we can ignore the reference and avoid emitting an unresolved reference.

Ideally, I think we should separate exports from reads. This could be part of a future PR.

Reply:
Considering what you said, I think it's best to have a function called `has_dual_binding` inside `Scope` and remove this function. It will be way easier to understand what we're doing and we can document that properly.